fixes #14453 - build and install Kafo parser cache

domcleal opened the pull request:
The parser cache will provide support for Puppet in AIO in the near
future with Kafo by not loading Puppet or puppet-strings, and provides
a performance boost to the installer startup. Requires Kafo 0.7.3.

Be careful to preserve mtimes of manifest files to ensure the cache
validity. Extra kafo-export-params search dirs were removed as all
current packages install correctly into /usr/bin or PATH, and it is
now required to build.

domcleal committed:
fixes #14453 - build and install Kafo parser cache

The parser cache will provide support for Puppet in AIO in the near
future with Kafo by not loading Puppet or puppet-strings, and provides
a performance boost to the installer startup. Requires Kafo 0.7.3.

Be careful to preserve mtimes of manifest files to ensure the cache
validity. Extra kafo-export-params search dirs were removed as all
current packages install correctly into /usr/bin or PATH, and it is
now required to build.

ekohl commented on Rakefile:
```diff
@@ -14,14 +14,23 @@ DATAROOTDIR = DATADIR = ENV['DATAROOTDIR'] || "#{PREFIX}/share"
 
+exporter = ENV['PATH'].split(':').push(
+  '/usr/bin',
+  ENV['KAFO_EXPORTER']).find do |exporter|
```
I don't like the reuse of exporter. Isn't dir (or something similar) a better naming?

domcleal replied:
Yep, fixed the variable name and I've split up the line to make it more readable.

mmoll commented:
[test]

domcleal commented:
We don't have any tests in this repo, unfortunately.  It's on my TODO list...

mmoll commented:
Oh, I won't get a DEB package to test out of here :)

domcleal commented:
Yeah, unfortunately not.  I did build an RPM with this patch, you could always inspect the contents with:

```
curl https://copr-be.cloud.fedoraproject.org/results/domcleal/foreman-puppet-aio/epel-7-x86_64/00175544-foreman-installer/foreman-installer-1.12.0-0.2.develop.git.0.bf22cb5.el7.centos.noarch.rpm | rpm2cpio - | cpio -di
```

or run `rake build install PREFIX=/tmp/installer` locally with kafo+librarian-puppet installed to build it.

mmoll commented:
this is looking good :+1:

mmoll commented:
please rebase ;)

mmoll commented:
the required kafo version in the Gemfile should also be bumped then

domcleal commented:
Thanks, it looks like there's an issue loading the serialised Puppet::\* classes from the cache, which I need to look at in more detail - moving to WoC for the time being.

domcleal commented:
http://projects.theforeman.org/issues/14729 describes the issue, which is an interaction between kafo 0.7.3 and kafo_parsers 0.1.0 when using this parser cache feature.  I think I'll wait until kafo 0.8.0 is ready as it should make the issue obsolete.

Edit: kafo 0.8.0 plus the next version of kafo_parsers.

domcleal commented:
Rebased and updated to require kafo 0.8.0, which causes `puppet` to get loaded again with the newer kafo_parsers.  The next version of kafo_parsers (probably out today) will also get rid of Puppet::\* from the cache entirely too.

mmoll commented:
merging, as we do have all needed fixes in the packaged versions of kafo and kafo_parsers now

mmoll commented:
merged, thanks @domcleal!
